### ISSUE TITLE
Gradient explosion of Seq2Seq (possibly related to linearlayer/embedding)

### ISSUE THREAD

**Issue by hikettei** [hikettei/clwaffe#101]
```lisp
(defmodel Encoder (vocab-size embedding-dim hidden-size)
  :parameters ((embedding (Embedding vocab-size embedding-dim :pad-idx 0))
               (layer     (RNN embedding-dim hidden-size :num-layers 1)))
  :forward ((x)
            (with-calling-layers x
                (embedding x)
                (layer x))))

(defmodel Decoder (vocab-size embedding-dim hidden-size)
  :parameters ((embedding (Embedding vocab-size embedding-dim :pad-idx 0))
               (layer     (RNN embedding-dim hidden-size :num-layers 1))
               (h2l       (linearlayer hidden-size vocab-size)))
  :forward ((encoder-state y)
            (let* ((ye (call (self embedding) y))
                   (hs (call (self layer) ye encoder-state))
                   (h-output (call (self h2l) hs)))
                  h-output)))

(defmodel Seq2Seq (vocab-size-lang1 vocab-size-lang2 embedding-dim input-size)
  :parameters ((encoder (Encoder vocab-size-lang1 embedding-dim input-size))
               (decoder (Decoder vocab-size-lang2 embedding-dim input-size)))
  :forward ((x y)
            (let ((x-state (call (self encoder) x)))
                 (call (self decoder)
                       x-state
                       (!aref y t '(0 -1))))))```

こんな感じのSeq2Seqがあって、

```lisp
(setq model (Seq2Seq 1000 1000 100 100))
```

vocab_sizeを1000以上で初期化して、学習を進めると2~3回で勾配爆発を起こす

LinearLayer/Embeddingの重みの初期値は適切だと思うんだけど、どこが悪いんだぁ〜ぐあぁ〜〜

**Comment by hikettei** [comment]
推定される原因：
LinearLayer: hidden-size -> vocab-size
Embedding

**Comment by hikettei** [comment]
cl-waffe initializes weights with Orthogonal/Xariver/He in response to their activation/model, but it won't work.
As I tried the same calculations on PyTorch (Seq2Seq with SimpleRNN), the value of loss is the same as cl-waffe, however optimizing is going well while cl-waffe is not.

**Comment by hikettei** [comment]
Hmm, i guess i need more tests, working on branch fix-rnn

**Comment by hikettei** [comment]
Fixed at #103, !softmax substracts x by the average of x but the axes was a different
